kore/src/Kore/Step/Axiom/Matcher.hs

```diff
@@ -636,34 +640,125 @@ rightAlignLists internal1 internal2
     -> Builtin.NormalizedAc normalized (TermLike Concrete) (TermLike variable)
     -> MaybeT (MatcherT variable unifier) ()
 matchNormalizedAc pushValue wrapTermLike normalized1 normalized2
-  | [] <- abstract2, [] <- opaque2
-  , [] <- abstract1
+  | [] <- excessAbstract2, [] <- opaque2
+  , [] <- excessAbstract1
   = do
-    Monad.guard (null excess1)
+    Monad.guard (null excessConcrete1)
     case opaque1 of
-        []       -> Monad.guard (null excess2)
+        []       -> do
+            Monad.guard (null excessConcrete2)
+            Monad.guard (null excessAbstract2)
         [frame1] -> push (Pair frame1 normalized2')
         _        -> empty
     Monad.Trans.lift $ Foldable.traverse_ pushValue concrete12
+    Monad.Trans.lift $ Foldable.traverse_ pushValue abstractMerge
   where
     normalized2' =
-        wrapTermLike normalized2 { Builtin.concreteElements = excess2 }
+        wrapTermLike normalized2
+            { Builtin.concreteElements = excessConcrete2
+            }
     abstract1 = Builtin.elementsWithVariables normalized1
     concrete1 = Builtin.concreteElements normalized1
     opaque1 = Builtin.opaque normalized1
     abstract2 = Builtin.elementsWithVariables normalized2
     concrete2 = Builtin.concreteElements normalized2
     opaque2 = Builtin.opaque normalized2
-    excess1 = Map.difference concrete1 concrete2
-    excess2 = Map.difference concrete2 concrete1
+
+    excessConcrete1 = Map.difference concrete1 concrete2
+    excessConcrete2 = Map.difference concrete2 concrete1
     concrete12 = Map.intersectionWith Pair concrete1 concrete2
+
+    IntersectionDifference
+        { intersection = abstractMerge
+        , excessFirst = excessAbstract1
+        , excessSecond = excessAbstract2
+        } = abstractIntersectionMerge abstract1 abstract2
+
+    abstractIntersectionMerge
+        :: [Builtin.Element normalized (TermLike variable)]
+        -> [Builtin.Element normalized (TermLike variable)]
+        -> IntersectionDifference
+            (Builtin.Element normalized (TermLike variable))
+            (Pair (Builtin.Value normalized (TermLike variable)))
+    abstractIntersectionMerge =
+        keyBasedIntersectionDifference elementKey elementMerger
+      where
+        elementKey
+            :: Builtin.Element normalized (TermLike variable)
+            -> TermLike variable
+        elementKey = fst . Builtin.unwrapElement
+        elementMerger
+            :: Builtin.Element normalized (TermLike variable)
+            -> Builtin.Element normalized (TermLike variable)
+            -> Pair (Builtin.Value normalized (TermLike variable))
+        elementMerger = Pair `on` (snd . Builtin.unwrapElement)
 matchNormalizedAc _ _ _ _ = empty
+
+data IntersectionDifference a b
+    = IntersectionDifference
+        { intersection :: ![b]
+        , excessFirst :: ![a]
+        , excessSecond :: ![a]
+        }
+
+keyBasedIntersectionDifference
```

Comment:
I think this function could be made simpler by taking two maps and using `Data.Align.align`:

```.hs
keyBasedIntersectionDifference
    :: forall a b k
    .  Ord k
    => (a -> a -> b)
    -> Map k a
    -> Map k a
    -> IntersectionDifference a b
keyBasedIntersectionDifference merger firsts seconds =
    foldl' helper emptyIntersectionDifference $ Map.elems $ Data.Align.align firsts seconds
  where
    helper accum@IntersectionDifference { excessFirst } (This a) =
        accum { excessFirst = a : excessFirst }
    helper accum@IntersectionDifference { excessSecond } (That a) =
        accum { excessSecond = a : excessSecond }
    helper accum@IntersectionDifference { intersection } (These first second) =
        accum { intersection = merger first second : intersection }
```

Reply:
Done.